webassembly/app.html: Fetch mesh from search-query url-param

jo-chemla opened the pull request:
PR Following feature request thread here https://github.com/f3d-app/f3d/issues/1595

snoyer commented on webassembly/app.html:
```diff
@@ -223,6 +221,27 @@ <h1 class="title">F3D Web</h1>
+          })
+          .then((buffer) => {
+            // Convert buffer to Uint8Array and openFile
+            Module.FS.writeFile('mesh_urlparam', new Uint8Array(buffer));
+            openFile('mesh_urlparam');
```
```suggestion
            const name = "..."; // retrieve name from mesh_url
            Module.FS.writeFile(name, new Uint8Array(buffer));
            openFile(name);
```
`f3d` needs the file extension to know the format and open the file

jo-chemla replied:
Ah that's right, is the extension parsed from the filename written to Module.FS? 
In that case, the mesh url has to end with the extension, and filename be retrieved from it, correct? 
Otherwise, we can as well add another url-param, like `loader` or `extension`, which if present will enforce loading with the corresponding loader, and if not present, will default to parsing extension from url. Would that make sense ?

snoyer replied:
> Ah that's right, is the extension parsed from the filename written to Module.FS?

yes this filename can be made up but it need the proper extension

> Otherwise, we can as well add another url-param, like `loader` or `extension`, which if present will enforce loading with the corresponding loader, and if not present, will default to parsing extension from url. Would that make sense ?

It does make sense, if the model was a glTF file but there was no way to tell from the url alone  (e.g. `http://example.com/foo/bar`) then you'd need some way to pass the format so that you could write the file as `bar.gltf` in the filesystem

jo-chemla replied:
Alright thanks for the precision that the file extension should be passed as filename. 
This extension can be enforced by user via url-param first, simply `extension` which would be appended to a placeholder name. 
At some point, one might try to extract it from the `Response Headers` as advised [here](https://github.com/kovacsv/Online3DViewer/issues/432), a bit trickier to parse, like `Content-Disposition: attachment; filename=odm_textured_model_geo.obj; filename*=UTF-8''odm_textured_model_geo.obj`

Meakk commented:
Just tried a local build and opening `http://192.168.0.12:3000/?mesh=https://github.com/KhronosGroup/glTF-Sample-Models/raw/main/2.0/DamagedHelmet/glTF-Binary/DamagedHelmet.glb` gives me the following error:
> Access to fetch at 'https://github.com/KhronosGroup/glTF-Sample-Models/raw/main/2.0/DamagedHelmet/glTF-Binary/DamagedHelmet.glb' from origin 'http://192.168.0.12:3000' has been blocked by CORS policy: The 'Access-Control-Allow-Origin' header contains the invalid value ''. Have the server send the header with a valid value, or, if an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.

Any idea?

jo-chemla commented:
True, CORS blocking is a standard error when fetching content that resides on a server, from an app that lives on another domain. You either need:
 - to use a url to a mesh file that has `Access-Control-Allow-Origin: *` headers, maybe [these](https://groups.csail.mit.edu/graphics/classes/6.837/F03/models/) should work like [teapot](https://groups.csail.mit.edu/graphics/classes/6.837/F03/models/teapot.obj)
 - or to use an Allow-CORS plugin (for example this plugin: [firefox](https://addons.mozilla.org/en-GB/firefox/addon/access-control-allow-origin/)/[chrome](https://chromewebstore.google.com/detail/allow-cors-access-control/lhobafahddgcelffkeicbaginigeejlf?pli=1))

Meakk commented:
Using `?mesh=https://groups.csail.mit.edu/graphics/classes/6.837/F03/models/teapot.obj`:
```
Uncaught (in promise) TypeError: Cannot read properties of null (reading 'split')
    at ?mesh=https://groups.csail.mit.edu/graphics/classes/6.837/F03/models/teapot.obj:241:45
```

Using `?mesh=https://groups.csail.mit.edu/graphics/classes/6.837/F03/models/teapot.obj&extension=.obj`:
No error is reported by the window is empty (a filename `meshurl_param.obj` is set in the open button)

![image](https://github.com/user-attachments/assets/95742d6c-6bea-4c91-80a5-220b2e8d60bd)

snoyer commented:
How do we feel about using the hash part of the URL (`https://f3d.app/web/#http://example.com/model.glb`) instead of the search params?
Also (and independently of the hash vs search question) appending the extension as a hash to the model URL looks like it could be a simpler solution than adding a seperate parameter

```js
        const mesh_url = window.location.hash.substring(1);
        const j = mesh_url.lastIndexOf('/');
        const mesh_name = (j>0? mesh_url.substring(j+1):mesh_url);
        const sanitized_mesh_name = mesh_name.replace(/[^a-zA-Z0-9._-]/,'_');
        // ...
            Module.FS.writeFile(sanitized_mesh_name, new Uint8Array(buffer));
            openFile(sanitized_mesh_name);
```

- with `http://localhost:8000/app.html#http://localhost:8000/Buggy.glb`:
  - `mesh_url`: `http://localhost:8000/Buggy.glb`
  - `mesh_name`: `Buggy.glb`
  - `sanitized_mesh_name`: `Buggy.glb`

- with `http://localhost:8000/app.html#http://localhost:8000/cube#.glb`:
  - `mesh_url`: `http://localhost:8000/cube#.glb`
  - `mesh_name`: `cube#.glb`
  - `sanitized_mesh_name`: `cube_.glb`

jo-chemla commented:
Final edit: 🚀 Last commit replaces the search-query with a hash as requested by @snoyer URLs are parsed exactly like search queries, but with a `#` rather than `?` character. mesh URLs parsed are decodeURI'ed, so one can pass either a standard url, or an encoded one in case it includes special characters like `?, # etc`. https://f3d.app/web#mesh=${encodeURI(https://groups.csail.mit.edu/graphics/classes/6.837/F03/models/teapot.obj)}&extension=.obj I've also updated the above list of mesh files I could test it against, and all formats do work! From my point-of-view, it is now up to the f3d team whether or not to merge!

Meakk commented on webassembly/app.html:
```diff
@@ -126,10 +126,12 @@ <h1 class="title">F3D Web</h1>
           const loader = Module.engineInstance.getLoader();
           if (loader.hasSceneReader(filePath))
           {
+            console.log('loader.hasSceneReader')
```
Please remove all `console.log` calls.

Meakk commented on webassembly/app.html:
```diff
@@ -223,6 +223,54 @@ <h1 class="title">F3D Web</h1>
+              throw new Error(`HTTP error, status = ${response.status}`);
+            }
+            // Build filename given extension urlparam or response header content-disposition
+            let filename 
```
Sorry, I have a C++ background, so let's pretend JS semicolon insertion doesn't exist and add a semicolon after each statement 😬

Meakk commented on webassembly/app.html:
```diff
@@ -223,6 +223,54 @@ <h1 class="title">F3D Web</h1>
+                  filename = mesh_url.split('/').pop()
+                  console.log('filename constructed from mesh_url filename+extension', filename)
+                  const url_extension = '.' + mesh_url.split('.').at(-1) 
+                  if (url_extension.length > 10) {
+                    throw new Error(`Bad extension parsed from filename ${filename}`)
+                  }
```
Is it necessary? If the extension is unknown, F3D will send an error message anyway.

Meakk commented on webassembly/app.html:
```diff
@@ -223,6 +223,54 @@ <h1 class="title">F3D Web</h1>
+                  if (url_extension.length > 10) {
+                    throw new Error(`Bad extension parsed from filename ${filename}`)
+                  }
+                  // filename = `mesh_urlparam.${url_extension}`
```
dead code to remove

Meakk reviewed (commented):
Thanks, I confirm it works, nice job :)  
Can you address the few cosmetic comments here, but it looks good to me.  
@snoyer can you also take a quick look at the changes?

jo-chemla commented:
Alright, just did include the suggested changes, thanks for the feedback! 
And I realized I forgot to push the `? -> #` change (search query to hash) suggested by snoyer, now it is included in this final commit.

snoyer commented on webassembly/app.html:
```diff
@@ -223,6 +221,49 @@ <h1 class="title">F3D Web</h1>
+        // Parse search-query mesh url-param or load default mesh file
+        // const params = new URLSearchParams(window.location.search);
+        // Replace first hash with question mark to have real search query parsing and avoid leading # in first parsed urlparam
+        const params = new URLSearchParams(window.location.hash.replace('#', '?'));
```
```suggestion
        const params = new URLSearchParams(window.location.hash.replace(/^#/, ''));
```
to remove only the `#` at the beginning

snoyer commented:
Nice feature! Here's a few improvements:

It'd be nice to clarify the order of priority used to get the filename between the download URL, response header, and `extension` parameter. I think it should be: from the header if available, otherwise from the url, and either way force the extension if the parameter is set. Maybe refactor to a `filename_for_model_url(model_url, format, header)` function so we can cover it with tests in the future.

Using the hash is nice and local, it doesn't get sent to the server and doesn't refresh the page when changed, but it therefore needs a listener to work on subsequent changes.

I'd say overall it would need to look something like:
```js
  function load_from_url(){
    const params = new URLSearchParams(window.location.hash.replace('#', '?'));
    // ...
    fetch(model_url).then((response) => {
         // ...
         const filename = filename_for_model_url(model_url, extension, response.headers.get("content-disposition")
         // ...
         openFile(filename);
    }
  }
  addEventListener("hashchange", (event) => {load_from_url()}); // do it whenever the hash changes
  load_from_url(); // do it the first time when page is loaded
```

jo-chemla commented:
Thanks, just refactored per your instructions, should be good to go! 

>  I think it should be: from the header if available, otherwise from the url, and either way force the extension if the parameter is set.

This is how it works at the moment - rephrased, if extension provided, enforce it, otherwise, get type from header if available, otherwise from url. 

```js
if (extension_parsed) {
  filename = `model.${extension_parsed}`;
} else {
  // If extension is not provided by user, try to get it auto from content-disposition header of url extension 
  if (contentDisposition) {
    filename = contentDisposition.split('filename=')[1].split(';')[0];
  } else {
    try {
      filename = model_url.split('/').pop();
```

jo-chemla commented on webassembly/app.html:
```diff
@@ -223,6 +221,57 @@ <h1 class="title">F3D Web</h1>
+          // Build filename given extension urlparam or response header content-disposition
+          let filename;
+          if (extension_parsed) {
+            filename = `model_urlparam.${extension_parsed}`;
```
alright, done!

snoyer commented on webassembly/app.html:
```diff
@@ -223,6 +221,57 @@ <h1 class="title">F3D Web</h1>
+              }
+            }
+          }
+          return filename;
```
oops, submitted the previous remark as a single comment instead of adding to a review, my bad!
```suggestion
          throw ...
```
You can throw here unconditionally, if we hit a good code path we would have returned already. That also ensure the function can never return an undefined `filename`

jo-chemla replied:
Ah true, just removed the try/catch and throw blocks. Thanks again!

snoyer replied:
`filename` doesn't exist anymore so you need to remove this last `return`

Meakk commented on webassembly/app.html:
```diff
@@ -223,6 +221,51 @@ <h1 class="title">F3D Web</h1>
+          // const params = new URLSearchParams(window.location.search);
+          // Replace first hash with question mark to have real search query parsing and avoid leading # in first parsed urlparam
+          const params = new URLSearchParams(window.location.hash.replace(/^#/, '?'));
+          const model_url_passed = params.get("mesh");
```
files are not only meshes, they can also contain materials and lights.  
Can we rename the option to `model` instead?

jo-chemla commented:
Thanks for the last suggestions, I included them and just pushed the final commit!

CharlesGueunet commented on webassembly/app.html:
```diff
@@ -223,6 +221,49 @@ <h1 class="title">F3D Web</h1>
         // do a first render and start the interactor
         Module.engineInstance.getWindow().render();
         Module.engineInstance.getInteractor().start();
+
```
Test
